src/bls12_381/ec.rs

```diff
@@ -166,6 +143,38 @@ macro_rules! curve_impl {
             fn into_projective(&self) -> $projective {
                 (*self).into()
             }
+
+            fn hash(t: Self::Base) -> Self
+            {
+                use ::LegendreSymbol::*;
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&Self::get_coeff_b());
+                w.add_assign(&Self::Base::one());
+                w = w.inverse().unwrap();
+                w.mul_assign(&Self::get_c1());
+                w.mul_assign(&t);
+
+                let mut x = w;
+                for i in 0..3 {
+                    match i {
+                        0 =>  { x.mul_assign(&t); x.negate(); x.add_assign(&Self::get_c2()) },
+                        1 =>  {                   x.negate(); x.sub_assign(&Self::Base::one()) },
+                        2 =>  { x=w; x.square();  x = x.inverse().unwrap(); x.add_assign(&Self::Base::one()) },
+                        _ =>  {}
+                    }
+
+                    if let Some(mut y) = Self::y2_from_x(x).sqrt() {
+                        if let QNonResidue = t.legendre() { y.negate() }
+                        return Self {x: x, y: y, infinity: false}
+                    }
+
+                }
+                Self {x: Self::Base::zero(), y: Self::Base::zero(), infinity: true }
```

Comment:
@ebfull, what's the policy for "cannot happen" cases? (Without having thought about it very hard, a ``panic!`` seems fine to me.)